banyan/src/store/mem_cache.rs

```diff
@@ -0,0 +1,123 @@
+use parking_lot::Mutex;
+use std::{convert::TryInto, hash::Hash, num::NonZeroUsize, sync::Arc, usize};
+use weight_cache::{Weighable, WeightCache};
+
+use super::{BlockWriter, ReadOnlyStore};
+
+/// Newtype wrapper for a boxed slice so we can implement Weighable
+#[derive(Debug)]
+struct MemBlock(Box<[u8]>);
+
+impl Weighable for MemBlock {
+    fn measure(value: &Self) -> usize {
+        value.0.len()
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Cache<L> {
+    cache: Arc<Mutex<WeightCache<L, MemBlock>>>,
+    /// maximum size of blocks to cache
+    /// we want this to remain very small, so we only cache tiny blocks
+    max_size: NonZeroUsize,
+}
+
+impl<L: Eq + Hash> Cache<L> {
+    fn new(max_size: usize, capacity: usize) -> Option<Self> {
+        let capacity = capacity.try_into().ok()?;
+        let max_size = max_size.try_into().ok()?;
+        Some(Self {
+            cache: Arc::new(Mutex::new(WeightCache::new(capacity))),
+            max_size,
+        })
+    }
+}
+
+/// A MemCache wraps an existing store with a cache with limited block size and limited total size
+///
+/// Note that this makes no attempts to remain consistent with the underlying store.
+/// If the underlying store is a content-addressed store, (links are hashes of data),
+/// you might get values despite the underlying store no longer having the data.
+///
+/// If it is not a content-addressed store, all bets are off and you might even get stale values.
+#[derive(Debug, Clone)]
+pub struct MemCache<L, I> {
+    inner: I,
+    /// the actual cache, None if a capacity of 0 was configured
+    cache: Option<Cache<L>>,
+}
+
+impl<L: Eq + Hash + Copy, I: ReadOnlyStore<L>> MemCache<L, I> {
+    /// create a new MemCache
+    /// `max_size` the maximum size for a block to be considered for caching, 0 to disable
+    /// `capacity` the total capacity of the cache, 0 to disable
+    pub fn new(inner: I, max_size: usize, capacity: usize) -> Self {
+        Self {
+            inner,
+            cache: Cache::new(max_size, capacity),
+        }
+    }
+
+    /// offer some data just to the cache, without writing it to the underlying store
+    pub fn offer(&self, link: &L, data: &[u8]) {
+        if let Some(cache) = self.cache.as_ref() {
+            if data.len() <= cache.max_size.into() {
+                let copy: Box<[u8]> = data.into();
```

Comment:
I still find `let copy = Box::<[u8]>::from(data)` clearer — `.into()` can be really opaque.

Reply:
I think in this case it is OK since the type is right there on the line. In general you are right - prefer from over into.